### 1. Wczytaj dane z pliku heart.csv i zapoznaj się z nimi. Wygeneruj statystyki.

In [15]:
import pandas as pd

pd.set_option('display.float_format', '{:0.2f}'.format)
data = pd.read_csv('heart.csv')

data.head()

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0   63    1   3     145   233    1        0       150     0     2.30    0   
1   37    1   2     130   250    0        1       187     0     3.50    0   
2   41    0   1     130   204    0        0       172     0     1.40    2   
3   56    1   1     120   236    0        1       178     0     0.80    2   
4   57    0   0     120   354    0        1       163     1     0.60    2   

   caa  thall  output  
0    0      1       1  
1    0      2       1  
2    0      2       1  
3    0      2       1  
4    0      2       1

In [16]:
data.describe()

age    sex     cp  trtbps   chol    fbs  restecg  thalachh   exng  \
count 303.00 303.00 303.00  303.00 303.00 303.00   303.00    303.00 303.00   
mean   54.37   0.68   0.97  131.62 246.26   0.15     0.53    149.65   0.33   
std     9.08   0.47   1.03   17.54  51.83   0.36     0.53     22.91   0.47   
min    29.00   0.00   0.00   94.00 126.00   0.00     0.00     71.00   0.00   
25%    47.50   0.00   0.00  120.00 211.00   0.00     0.00    133.50   0.00   
50%    55.00   1.00   1.00  130.00 240.00   0.00     1.00    153.00   0.00   
75%    61.00   1.00   2.00  140.00 274.50   0.00     1.00    166.00   1.00   
max    77.00   1.00   3.00  200.00 564.00   1.00     2.00    202.00   1.00   

       oldpeak    slp    caa  thall  output  
count   303.00 303.00 303.00 303.00  303.00  
mean      1.04   1.40   0.73   2.31    0.54  
std       1.16   0.62   1.02   0.61    0.50  
min       0.00   0.00   0.00   0.00    0.00  
25%       0.00   1.00   0.00   2.00    0.00  
50%       0.80   1.00   0.00   2.00    1.00  
75%       1.60   2.00   1.00   3.00    1.00  
max       6.20   2.00   4.00   3.00    1.00

In [17]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trtbps    303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalachh  303 non-null    int64  
 8   exng      303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slp       303 non-null    int64  
 11  caa       303 non-null    int64  
 12  thall     303 non-null    int64  
 13  output    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [18]:
data.isna().sum()

age         0
sex         0
cp          0
trtbps      0
chol        0
fbs         0
restecg     0
thalachh    0
exng        0
oldpeak     0
slp         0
caa         0
thall       0
output      0
dtype: int64

### 2. Całkowicie losowo wygeneruj brakujące dane tak, aby w kolumnach age i chol brakowało po około 10% wartości.


In [19]:
import numpy as np
import random 
rand_data = data.copy()

missing = 0.1
columns = ['age', 'chol']

for col in columns:
    i = 0
    while i < len(rand_data):
        p = random.random()
        if(p<=missing+0.0001):
            rand_data.loc[i, col] = np.nan
        i+=1

In [20]:
rand_data.isnull().sum()

age         35
sex          0
cp           0
trtbps       0
chol        36
fbs          0
restecg      0
thalachh     0
exng         0
oldpeak      0
slp          0
caa          0
thall        0
output       0
dtype: int64

In [21]:
rand_data.head(10)

age  sex  cp  trtbps   chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0 63.00    1   3     145 233.00    1        0       150     0     2.30    0   
1 37.00    1   2     130 250.00    0        1       187     0     3.50    0   
2 41.00    0   1     130 204.00    0        0       172     0     1.40    2   
3 56.00    1   1     120 236.00    0        1       178     0     0.80    2   
4 57.00    0   0     120 354.00    0        1       163     1     0.60    2   
5 57.00    1   0     140 192.00    0        1       148     0     0.40    1   
6 56.00    0   1     140 294.00    0        0       153     0     1.30    1   
7 44.00    1   1     120    NaN    0        1       173     0     0.00    2   
8 52.00    1   2     172 199.00    1        1       162     0     0.50    2   
9 57.00    1   2     150 168.00    0        1       174     0     1.60    2   

   caa  thall  output  
0    0      1       1  
1    0      2       1  
2    0      2       1  
3    0      2       1  
4    0      2       1  
5    0      1       1  
6    0      2       1  
7    0      3       1  
8    0      3       1  
9    0      2       1

### 3. Usuń rekordy zawierające jakiekolwiek brakujące wartości.

In [22]:
drop_rand_data = rand_data.dropna()

print(len(drop_rand_data))

236


In [23]:
drop_rand_data.isnull().sum()

age         0
sex         0
cp          0
trtbps      0
chol        0
fbs         0
restecg     0
thalachh    0
exng        0
oldpeak     0
slp         0
caa         0
thall       0
output      0
dtype: int64

In [24]:
drop_rand_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 236 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       236 non-null    float64
 1   sex       236 non-null    int64  
 2   cp        236 non-null    int64  
 3   trtbps    236 non-null    int64  
 4   chol      236 non-null    float64
 5   fbs       236 non-null    int64  
 6   restecg   236 non-null    int64  
 7   thalachh  236 non-null    int64  
 8   exng      236 non-null    int64  
 9   oldpeak   236 non-null    float64
 10  slp       236 non-null    int64  
 11  caa       236 non-null    int64  
 12  thall     236 non-null    int64  
 13  output    236 non-null    int64  
dtypes: float64(3), int64(11)
memory usage: 27.7 KB


### 4. Dokonaj imputacji metodą średniej.

In [ ]:
mean_imputation = rand_data.fillna(rand_data.mean()) # wstawiamy średnia w miejsce NaN

mean_imputation.describe()

age    sex     cp  trtbps   chol    fbs  restecg  thalachh   exng  \
count 303.00 303.00 303.00  303.00 303.00 303.00   303.00    303.00 303.00   
mean   54.64   0.68   0.97  131.62 246.50   0.15     0.53    149.65   0.33   
std     8.58   0.47   1.03   17.54  49.44   0.36     0.53     22.91   0.47   
min    29.00   0.00   0.00   94.00 126.00   0.00     0.00     71.00   0.00   
25%    50.00   0.00   0.00  120.00 214.00   0.00     0.00    133.50   0.00   
50%    54.64   1.00   1.00  130.00 246.50   0.00     1.00    153.00   0.00   
75%    60.00   1.00   2.00  140.00 269.00   0.00     1.00    166.00   1.00   
max    77.00   1.00   3.00  200.00 564.00   1.00     2.00    202.00   1.00   

       oldpeak    slp    caa  thall  output  
count   303.00 303.00 303.00 303.00  303.00  
mean      1.04   1.40   0.73   2.31    0.54  
std       1.16   0.62   1.02   0.61    0.50  
min       0.00   0.00   0.00   0.00    0.00  
25%       0.00   1.00   0.00   2.00    0.00  
50%       0.80   1.00   0.00   2.00    1.00  
75%       1.60   2.00   1.00   3.00    1.00  
max       6.20   2.00   4.00   3.00    1.00

In [ ]:
mean_imputation.head(10)

age  sex  cp  trtbps   chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0 63.00    1   3     145 233.00    1        0       150     0     2.30    0   
1 37.00    1   2     130 250.00    0        1       187     0     3.50    0   
2 41.00    0   1     130 204.00    0        0       172     0     1.40    2   
3 56.00    1   1     120 236.00    0        1       178     0     0.80    2   
4 57.00    0   0     120 354.00    0        1       163     1     0.60    2   
5 57.00    1   0     140 192.00    0        1       148     0     0.40    1   
6 56.00    0   1     140 294.00    0        0       153     0     1.30    1   
7 44.00    1   1     120 246.50    0        1       173     0     0.00    2   
8 52.00    1   2     172 199.00    1        1       162     0     0.50    2   
9 57.00    1   2     150 168.00    0        1       174     0     1.60    2   

   caa  thall  output  
0    0      1       1  
1    0      2       1  
2    0      2       1  
3    0      2       1  
4    0      2       1  
5    0      1       1  
6    0      2       1  
7    0      3       1  
8    0      3       1  
9    0      2       1

### 5. Porównaj otrzymane rezultaty z oryginalnymi danymi.


In [ ]:
# data['age'].compare(mean_imputation['age'])
data['chol'].compare(mean_imputation['chol'])

self  other
7   263.00 246.50
11  275.00 246.50
17  226.00 246.50
23  243.00 246.50
27  175.00 246.50
29  197.00 246.50
43  264.00 246.50
44  321.00 246.50
45  325.00 246.50
49  234.00 246.50
89  248.00 246.50
95  226.00 246.50
98  315.00 246.50
101 270.00 246.50
113 211.00 246.50
118 204.00 246.50
132 295.00 246.50
136 178.00 246.50
139 263.00 246.50
140 295.00 246.50
159 221.00 246.50
164 175.00 246.50
167 268.00 246.50
171 229.00 246.50
174 206.00 246.50
193 282.00 246.50
194 185.00 246.50
204 164.00 246.50
211 260.00 246.50
218 254.00 246.50
234 322.00 246.50
250 298.00 246.50
274 275.00 246.50
276 218.00 246.50
292 225.00 246.50
300 193.00 246.50

### 6. Dokonaj imputacji metodami uzupełniania braków poprzednią wartością, interpolacji, MICE i k-najbliższych sąsiadów.


#### *Uzupełnianie poprzednią wartością (Forward Fill)*

In [37]:
forward_fill = rand_data.copy().ffill().dropna() # (jesli np. w pierwszym rekordzie bedzie brak to forward fill zostawi NaN, nie wypelni go)

forward_fill.describe()

age    sex     cp  trtbps   chol    fbs  restecg  thalachh   exng  \
count 303.00 303.00 303.00  303.00 303.00 303.00   303.00    303.00 303.00   
mean   54.74   0.68   0.97  131.62 245.95   0.15     0.53    149.65   0.33   
std     9.24   0.47   1.03   17.54  52.58   0.36     0.53     22.91   0.47   
min    29.00   0.00   0.00   94.00 126.00   0.00     0.00     71.00   0.00   
25%    47.50   0.00   0.00  120.00 211.00   0.00     0.00    133.50   0.00   
50%    56.00   1.00   1.00  130.00 239.00   0.00     1.00    153.00   0.00   
75%    62.00   1.00   2.00  140.00 273.50   0.00     1.00    166.00   1.00   
max    77.00   1.00   3.00  200.00 564.00   1.00     2.00    202.00   1.00   

       oldpeak    slp    caa  thall  output  
count   303.00 303.00 303.00 303.00  303.00  
mean      1.04   1.40   0.73   2.31    0.54  
std       1.16   0.62   1.02   0.61    0.50  
min       0.00   0.00   0.00   0.00    0.00  
25%       0.00   1.00   0.00   2.00    0.00  
50%       0.80   1.00   0.00   2.00    1.00  
75%       1.60   2.00   1.00   3.00    1.00  
max       6.20   2.00   4.00   3.00    1.00

#### *Uzupełnianie metodami interpolacji*

In [44]:
interpolation_fill = rand_data.fillna(rand_data.interpolate(method = 'linear'))

interpolation_fill.describe()

age    sex     cp  trtbps   chol    fbs  restecg  thalachh   exng  \
count 303.00 303.00 303.00  303.00 303.00 303.00   303.00    303.00 303.00   
mean   54.80   0.68   0.97  131.62 246.61   0.15     0.53    149.65   0.33   
std     8.96   0.47   1.03   17.54  50.46   0.36     0.53     22.91   0.47   
min    29.00   0.00   0.00   94.00 126.00   0.00     0.00     71.00   0.00   
25%    48.50   0.00   0.00  120.00 212.75   0.00     0.00    133.50   0.00   
50%    56.00   1.00   1.00  130.00 240.00   0.00     1.00    153.00   0.00   
75%    61.25   1.00   2.00  140.00 273.50   0.00     1.00    166.00   1.00   
max    77.00   1.00   3.00  200.00 564.00   1.00     2.00    202.00   1.00   

       oldpeak    slp    caa  thall  output  
count   303.00 303.00 303.00 303.00  303.00  
mean      1.04   1.40   0.73   2.31    0.54  
std       1.16   0.62   1.02   0.61    0.50  
min       0.00   0.00   0.00   0.00    0.00  
25%       0.00   1.00   0.00   2.00    0.00  
50%       0.80   1.00   0.00   2.00    1.00  
75%       1.60   2.00   1.00   3.00    1.00  
max       6.20   2.00   4.00   3.00    1.00

In [ ]:
interpolation_fill2 = rand_data.fillna(rand_data.interpolate(method = 'polynomial', order = 2))

interpolation_fill2.describe()

age    sex     cp  trtbps   chol    fbs  restecg  thalachh   exng  \
count 303.00 303.00 303.00  303.00 303.00 303.00   303.00    303.00 303.00   
mean   54.81   0.68   0.97  131.62 247.21   0.15     0.53    149.65   0.33   
std     9.61   0.47   1.03   17.54  54.92   0.36     0.53     22.91   0.47   
min    25.02   0.00   0.00   94.00 126.00   0.00     0.00     71.00   0.00   
25%    48.00   0.00   0.00  120.00 208.50   0.00     0.00    133.50   0.00   
50%    56.00   1.00   1.00  130.00 241.00   0.00     1.00    153.00   0.00   
75%    62.00   1.00   2.00  140.00 277.00   0.00     1.00    166.00   1.00   
max    79.09   1.00   3.00  200.00 564.00   1.00     2.00    202.00   1.00   

       oldpeak    slp    caa  thall  output  
count   303.00 303.00 303.00 303.00  303.00  
mean      1.04   1.40   0.73   2.31    0.54  
std       1.16   0.62   1.02   0.61    0.50  
min       0.00   0.00   0.00   0.00    0.00  
25%       0.00   1.00   0.00   2.00    0.00  
50%       0.80   1.00   0.00   2.00    1.00  
75%       1.60   2.00   1.00   3.00    1.00  
max       6.20   2.00   4.00   3.00    1.00

### 7. Dla każdej z metod policz najmniejszy i największy błąd, a także miary MAE i RMSE.

In [ ]:
# Zrobic też

### 8. Utwórz wykres pokazujący zależność między poziomem cholesterolu a wiekiem pacjenta dla danych oryginalnych i po przeprowadzonych imputacjach.

In [ ]:
# Zrobic wykry dla kazdej imputacji i dla oryginalnych danych

### 9. Przeprowadź analogiczne analizy dla 30% brakujących wartości.

In [ ]:
# Wystarczy zmienic 0.1 na 0.3, nic nie przepisywac

### 10. Zbadaj hipotezę, że średni poziom cholesterolu wynosi 250.

### 11. Sprawdź, czy na podstawie poziomu cholesterolu jesteś w stanie rozpoznać płeć pacjenta.